agent/api/task/task.go

```diff
@@ -1213,3 +1219,19 @@ func (task *Task) AddResource(resourceType string, resource taskresource.TaskRes
 	defer task.lock.Unlock()
 	task.ResourcesMapUnsafe[resourceType] = append(task.ResourcesMapUnsafe[resourceType], resource)
 }
+
+// SetTerminalReason sets the terminalReason string and this can only be set
+// once per the task's lifecycle. This field does not accept updates.
+func (task *Task) SetTerminalReason(reason string) {
+	task.terminalReasonOnce.Do(func() {
```

Comment:
An Info log here stating that you're trying to set a task's terminal reason would be helpful in this first one wins scenario

Reply:
this is a good idea, adding.